### PR TITLE
TraceLens GPU benchmarking for Missing MAF Values

### PR DESCRIPTION
**Pull request by mohbasit**
# PR #491 — Auto-generate TraceLens GPU arch JSON via microbenchmark

## Summary

TraceLens roofline analysis needs a per-GPU architecture spec (`TraceLens/Agent/Analysis/utils/arch/<platform>.json`). Today that file must already exist (bundled with TraceLens or checked in); if it's missing for a platform like MI355X, the kernel-agent TraceLens pipeline does not produce roofline metrics.

This PR adds an automatic **pre-report GPU arch benchmark** step to `tracelens_analysis.py`:

- **New module** `tracelens_arch_benchmark.py` resolves the bundled arch JSON for `--target-platform` when present.
- **When missing**, it runs `TraceLens.PerfModel.benchmarking.microbench` to measure and write the spec before trace splitting / report generation.
- **GPU selection** picks an idle visible GPU via TraceLens `check_gpu_idle`, then pins `HIP_VISIBLE_DEVICES` / `CUDA_VISIBLE_DEVICES` / `ROCR_VISIBLE_DEVICES` **only in the microbenchmark subprocess** (parent process env unchanged).
- **Microbenchmark settings**: `--warmup 20 --rep 50`.
- **Timeout** is configurable via `TRACELENS_ARCH_BENCHMARK_TIMEOUT_SEC` (minimum 600s).
- The resolved path is recorded as artifact `tracelens_gpu_arch_json`.

Supporting changes in `tracelens_analysis.py`:

- Calls `populate_gpu_arch_json` as a new pipeline step (`populate_gpu_arch_json`) before trace splitting.
- Extends `run_command()` with an optional `env=` argument for subprocess-scoped GPU pinning.
- Hoists lazy imports (`csv`, `aiter`, framework patch roots) to module top level for clarity.

Unit tests in `test_tracelens_arch_benchmark.py` cover platform normalization, GPU candidate discovery, idle-GPU selection, bundled-spec short-circuit, and microbenchmark invocation when the spec is missing.

## Test plan

- [ ] `pytest kernel-agent/tools/test_tracelens_arch_benchmark.py`
- [ ] Run `tracelens_analysis.py` with a platform that has a **bundled** arch JSON (e.g. MI300X) — confirm log shows `using bundled spec` and no microbenchmark runs
- [ ] Run with a platform **without** bundled spec (e.g. MI355X) on a node with a free GPU with ONLY open source TraceLens (no Tracelens-internal) — confirm microbenchmark runs, writes `arch/<platform>.json`, and pipeline continues to trace split/report
- [ ] Run when all visible GPUs are busy — confirm clear `no unoccupied GPU` error
- [ ] Verify `TRACELENS_ARCH_BENCHMARK_TIMEOUT_SEC` is honored and status JSON shows step `populate_gpu_arch_json`

### CODE REVIEW

File: kernel-agent/tools/test_tracelens_arch_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+import json
+import os
+import sys
+from pathlib import Path
+from unittest.mock import patch
+
+import pytest
+
+TOOLS_DIR = Path(__file__).resolve().parent
```
Comment (tsrikris) [inline kernel-agent/tools/test_tracelens_arch_benchmark.py]: Align all imports to the top followed by env vars

File: kernel-agent/tools/test_tracelens_arch_benchmark.py
Comment (tsrikris) [inline kernel-agent/tools/test_tracelens_arch_benchmark.py:1]: Is this test suite better placed within TraceLens?

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+from __future__ import annotations
+
+import json
+import os
+import sys
+from pathlib import Path
+from typing import Callable
+
+
+RunCommandFn = Callable[..., int]
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py]: Are these really needed? Only a single use

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    "CUDA_VISIBLE_DEVICES",
+    "ROCR_VISIBLE_DEVICES",
+)
+
+
+def normalize_platform(platform: str) -> str:
+    return (platform or "").strip().upper()
+
+
+def count_visible_gpus() -> int:
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py]: Why is this needed?

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            continue
+        parts = [part.strip() for part in val.split(",") if part.strip()]
+        if not parts:
+            continue
+        try:
+            return [int(part) for part in parts]
+        except ValueError:
+            return list(range(len(parts)))
+    try:
+        import torch
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py]: Move imports to the top

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+        if torch.cuda.is_available():
+            return list(range(int(torch.cuda.device_count())))
+    except Exception:
+        pass
+    return []
+
+
+def pin_single_physical_gpu(physical_id: int) -> None:
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py]: Does this affect other downstream steps or only this one?

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    """Expose exactly one physical GPU to downstream CUDA/HIP consumers."""
+    value = str(physical_id)
+    os.environ["HIP_VISIBLE_DEVICES"] = value
+    os.environ["CUDA_VISIBLE_DEVICES"] = value
+    os.environ["ROCR_VISIBLE_DEVICES"] = value
+
+
+def select_idle_gpu(*, log: LogFn | None = None, util_threshold: int = 5) -> int:
+    """Pick an unoccupied GPU and pin it as the only visible device."""
+    from TraceLens.PerfModel.benchmarking.microbench_utils import check_gpu_idle
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py]: Move imports to the top

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+    raise RuntimeError(
+        "gpu_arch_benchmark found no unoccupied GPU among "
+        f"{candidates}. {'; '.join(busy_reports)}"
+    )
+
+
+def resolve_arch_json_path(platform: str) -> Path | None:
+    """Return the bundled arch JSON path for ``platform``, if present."""
+    from TraceLens.Agent.Analysis.utils.arch_utils import _collect_arch_jsons
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py]: Imports at the top

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    if not canonical:
+        return None
+    for name, path in _collect_arch_jsons().items():
+        if name.upper() == canonical:
+            return Path(path)
+    return None
+
+
+def default_arch_output_path(tracelens_root: Path, platform: str) -> Path:
+    canonical = normalize_platform(platform)
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py:134]: Function needed? Why not in-line single use function

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    return (
+        tracelens_root
+        / "TraceLens/Agent/Analysis/utils/arch"
+        / f"{canonical}.json"
+    )
+
+
+MICROBENCH_WARMUP = 20
+MICROBENCH_REP = 50
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py]: Move to the top

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+    out_path = default_arch_output_path(tracelens_root, canonical)
+    log(
+        "gpu_arch_json: no bundled spec for "
+        f"{canonical}; running TraceLens microbenchmark -> {out_path}"
+    )
+
+    select_idle_gpu(log=log)
+
+    rc = run_command(
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py:170]: Is this the approach other scripts in the repo use?

File: kernel-agent/tools/tracelens_analysis.py [hunk 11]
```diff
@@ -2903,6 +2823,38 @@ def main() -> int:
...
+                log=lambda msg: append_log(log_path, msg),
+                run_command=lambda cmd, *, cwd, timeout_s: run_command(
+                    cmd,
+                    cwd=cwd,
+                    log_path=log_path,
+                    timeout_s=timeout_s,
+                ),
+                timeout_s=arch_benchmark_timeout_s,
+            )
+            artifacts["tracelens_gpu_arch_json"] = str(gpu_arch_path)
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_analysis.py:2874]: Is this wired into the rest of the agent flow? Meaning does any subsequent step handle things correctly? The future agent calls and the rooflining scripts towards the end

File: kernel-agent/tools/tracelens_arch_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        / "TraceLens/Agent/Analysis/utils/arch"
+        / f"{canonical}.json"
+    )
+
+
+MICROBENCH_WARMUP = 20
+MICROBENCH_REP = 50
+
+
+def ensure_gpu_arch_json(
```
Comment (tsrikris) [inline kernel-agent/tools/tracelens_arch_benchmark.py]: Can the naming of the function be changed? ensure seems more like a check vs this function which handles missing MAF values.

File: kernel-agent/tools/tracelens_analysis.py
Comment (tsrikris) [inline kernel-agent/tools/tracelens_analysis.py:1]: Can many of these functions be re-orged into a single class (confirm coding pattern in rest of hyperloom). If the only consumer of the functions here are in this file, then it might be cleaner to re-org into a class with private methods

File: kernel-agent/tools/test_tracelens_arch_benchmark.py
Comment (tsrikris) [reply kernel-agent/tools/test_tracelens_arch_benchmark.py:1]: Would recommend only adding short tests if absolutely needed on the Hyperloom specific portions

**Review by tsrikris — DISMISSED** [review]
Added few comments for review @mohbasit
